envoy/config/filter/http/ext_authz/v2/ext_authz.proto

```diff
@@ -23,4 +23,25 @@ message ExtAuthz {
   // is NOT denied then traffic will be permitted.
   // Defaults to false.
   bool failure_mode_allow = 2;
+
+  // When this value is set to true, the filter must call an HTTP authorization service only.
+  bool use_http_service = 3;
+
+  // The external authorization HTTP service configuration. Only applicable when `use_http_service`
+  // is true.
+  message HttpService {
+
+    // Sets the cluster name which the authorization request must be sent to.
+    string cluster = 1;
+
+    // Sets the time in milliseconds within the service should respond to an authorization request.
+    uint32 timeout_ms = 2;
+
+    // Sets a list of authorization response headers that are allowed to be injected in the
+    // downstream client request before dispatching it to the upstream.
+    repeated string allowed_headers = 3;
```

Comment:
Why is this necessary? Don't we trust the authz server to do the right thing?

Reply:
+1

Reply:
I just copied this from the Datawire's filter. It seems to prevent that all authz headers get copied over to the original request. E.g., content-length.
```
  if (!config_->allowed_headers_.empty()) {
    // Yup. Let's see if any of them are present.

    for (const std::string& allowed_header : config_->allowed_headers_) {
      LowerCaseString key(allowed_header);

      // OK, do we have this header?

      const HeaderEntry* hdr = response->headers().get(key);

      if (hdr) {
        // Well, the header exists at all. Does it have a value?

        const HeaderString& value = hdr->value();

        if (!value.empty()) {
          // Not empty! Copy it into our request_headers_.

          std::string valstr{value.c_str()};

          ENVOY_STREAM_LOG(trace, "ExtAuth allowing response header {}: {}", *callbacks_,
                           allowed_header, valstr);
          addedHeaders = true;
          request_headers_->addCopy(key, valstr);
        }
      }
    }
```
https://github.com/datawire/envoy/blob/546f6bbe3b74249d871073ef514cf71144e88dbf/source/common/http/filter/extauth.cc#L215

@kflynn Would you mind to help us with this one? Thanks!

Reply:
@htuch @mattklein123 Folks we've talked to have been less comfortable with given an auth service carte blanche to change anything it wants going upstream, more comfortable with only allowing it to change known things (cf some of the comments around trust on https://docs.google.com/document/d/1L3aRk9yaT6Lsb4epI6rdjCdStOuwTfhoeSP82zunkGg/edit).

Reply:
I'm not sure I really buy this. Why would the filter then trust anything the auth service does? Whether the decision, the response, etc.?

Reply:
@mattklein123 I think there were three underlying concerns:

- Bugs in the auth service are always a thing; having some protection in the upstream-modification path is a bit of a safeguard. Same idea as not wanting even your favorite Unix developer run everything as root.

- Policy enforcement is sometimes a thing. It may be necessary to be able to tell auditors that only this _one specific set_ of modifications can ever happen, unless the following audited process is followed, etc.

- Finally, sure, malicious auth services might someday be a thing somewhere? but they clearly do not need unchecked upstream-header modification to wreak havoc, so I don't consider protecting against that case to be a major driver for this config element.

Reply:
I agree that it seems a bit arbitrary just policing the headers. If we don't trust the auth service, it can do a lot bad things, e.g. inject content in the reply for the body.

I can see that in some applications, you do want to strip headers unconditionally, when something passes between internal <-> external on edge proxies. We could use a separate filter in the stack for this, would that work in your application? It seems if the auth server can't set that header, backends probably shouldn't be trusted to do it either?

Reply:
I'm not sure, it seems to be the same problem.
Would be better to just enforce a contract in the same way that it's done with gRPC?  

200 OK
```
{
   "headers": [
       "custom-header" : "123",
       "custom-header2" : "456"
   ]
}
```

401 NOT_AUTHORIZED
```
{
   "status_code": 301,
   "headers": [
       "custom-header" : "123",
       "custom-header2" : "456"
   ],
   "body": "redirecting..."
}
```